### DISCUSSION TITLE
Trying to build Lightspark for Windows

### DISCUSSION THREAD

**Discussion by n0samu** [lightspark/lightspark#604] [Q&A] [answered]
I am trying to build for Windows (64-bit) using Ubuntu 20.04 on WSL1 on Windows 10. This is a fresh installation and I updated all packages.
I followed the instructions here: https://github.com/lightspark/lightspark/blob/master/README.win32
The installed CMake version is 3.16.3 and it is using i686-w64-mingw32.static-gcc 5.5.0.
After installing the dependencies and MXE, I cloned the lightspark repo and ran this command: 
```
cmake -DCMAKE_INSTALL_PREFIX=../mingw-build -DCMAKE_BUILD_TYPE=Debug -G'Unix Makefiles' -DCMAKE_TOOLCHAIN_FILE=../mxe/usr/i686-w64-mingw32.static/share/cmake/mxe-conf.cmake ../lightspark
```
Then, after running `make package`, I got these errors:
```
/home/nosamu/lightspark/src/backends/input.cpp: In member function 'bool lightspark::InputThread::handleKeyboardShortcuts(const SDL_KeyboardEvent*)':
/home/nosamu/lightspark/src/backends/input.cpp:434:51: error: no matching function for call to 'lightspark::EngineData::setDisplayState(const char [7])'
   m_sys->getEngineData()->setDisplayState("normal");
                                                   ^
In file included from /home/nosamu/lightspark/src/scripting/flash/display3d/flashdisplay3d.h:28:0,
                 from /home/nosamu/lightspark/src/scripting/flash/display/flashdisplay.h:33,
                 from /home/nosamu/lightspark/src/parsing/tags.h:29,
                 from /home/nosamu/lightspark/src/scripting/abc.h:25,
                 from /home/nosamu/lightspark/src/backends/input.cpp:20:
/home/nosamu/lightspark/src/platforms/engineutils.h:149:15: note: candidate: virtual void lightspark::EngineData::setDisplayState(const lightspark::tiny_string&, lightspark::SystemState*)
  virtual void setDisplayState(const tiny_string& displaystate, SystemState *sys);
               ^
/home/nosamu/lightspark/src/platforms/engineutils.h:149:15: note:   candidate expects 2 arguments, 1 provided
/home/nosamu/lightspark/src/backends/input.cpp:468:112: error: no matching function for call to 'lightspark::EngineData::setDisplayState(const char*)'
    m_sys->getEngineData()->setDisplayState(m_sys->getEngineData()->inFullScreenMode() ? "normal" : "fullScreen");
                                                                                                                ^
In file included from /home/nosamu/lightspark/src/scripting/flash/display3d/flashdisplay3d.h:28:0,
                 from /home/nosamu/lightspark/src/scripting/flash/display/flashdisplay.h:33,
                 from /home/nosamu/lightspark/src/parsing/tags.h:29,
                 from /home/nosamu/lightspark/src/scripting/abc.h:25,
                 from /home/nosamu/lightspark/src/backends/input.cpp:20:
/home/nosamu/lightspark/src/platforms/engineutils.h:149:15: note: candidate: virtual void lightspark::EngineData::setDisplayState(const lightspark::tiny_string&, lightspark::SystemState*)
  virtual void setDisplayState(const tiny_string& displaystate, SystemState *sys);
               ^
/home/nosamu/lightspark/src/platforms/engineutils.h:149:15: note:   candidate expects 2 arguments, 1 provided
make[2]: *** [src/CMakeFiles/spark.dir/build.make:344: src/CMakeFiles/spark.dir/backends/input.cpp.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:221: src/CMakeFiles/spark.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
```
What can I do to diagnose this issue?

**Comment by dbluelle** [comment, marked as answer]
Sorry, that was my fault, one changed file was not included in the last patch, it should work now.

**Reply by n0samu** [reply]
Thank you for the quick response! It successfully builds now, but when I run lightspark.exe, it immediately crashes (it doesn't print anything or show a window). Using Process Monitor, I found that lightspark.exe crashes right after failing to find files on these paths:
```
C:\etc\gnutls\config
C:\etc\gnutls\
C:\home\nosamu\mxe\usr\i686-w64-mingw32.static\share\locale\en_US\LC_MESSAGES\messages.mo
C:\home\nosamu\mxe\usr\i686-w64-mingw32.static\share\locale\en\LC_MESSAGES\messages.mo
```
I guess I need to change these paths somehow, but how would I do that?

Also, the NSIS installer fails to build. Here is the error in the log:
```
Generating language tables... Unable to convert processed string "فضای موجود: " to codepage 1256
makensis: Source/strlist.cpp:131: unsigned int ExeHeadStringList::find(const void*, unsigned int, WORD, bool, char**) const: Assertion `cbMB' failed.
Child aborted
```
I can't find where that string is or what exactly is wrong with it.

**Reply by dbluelle** [reply]
Hmm there seems to be something wrong with the gettext step.
However, the instructions in README.win32 seem to be quite outdated, I don't know if those still work.
I normally use the cmake included in mxe to avoid any cross-compiling problems:
```
export PATH=/path/to/mxe/usr/bin:$PATH
x86_64-w64-mingw32.static-cmake ../lightspark
make
```
btw I think we should remove the gettext stuff, as it's unmaintained and only used to translate some debug output messages

**Reply by eudoxos** [reply]
> Also, the NSIS installer fails to build. Here is the error in the log:
> 
> ```
> Generating language tables... Unable to convert processed string "فضای موجود: " to codepage 1256
> makensis: Source/strlist.cpp:131: unsigned int ExeHeadStringList::find(const void*, unsigned int, WORD, bool, char**) const: Assertion `cbMB' failed.
> Child aborted
> ```
> 
> I can't find where that string is or what exactly is wrong with it.

I was hitting this exact same error (packaging a different piece of software) and the workaround was to run nsis with LC_ALL set to non-utf8 locale (I used `export LC_ALL=C`).
